src/metrics/macros.rs

```diff
@@ -0,0 +1,67 @@
+pub(super) trait MetricFromOpts: Sized {
+    fn from_opts(opts: prometheus::Opts) -> Result<Self, prometheus::Error>;
+}
+
+#[macro_export]
+macro_rules! metrics {
+    (
+        $vis:vis struct $name:ident {
+            $(
+                #[doc = $help:expr]
+                $metric_vis:vis $metric:ident: $ty:ty $([$($label:expr),* $(,)?])?
+            ),* $(,)?
+        }
+        namespace: $namespace:expr,
+    ) => {
+        $vis struct $name {
+            registry: prometheus::Registry,
+            $($metric_vis $metric: $ty,)*
+        }
+        impl $name {
+            $vis fn new() -> Result<Self, prometheus::Error> {
+                let registry = prometheus::Registry::new();
+                $(
+                    let $metric = <$ty>::from_opts(
+                        prometheus::Opts::new(stringify!($metric), $help)
+                            .namespace($namespace)
+                            $(.variable_labels(vec![$($label.into()),*]))?
+                    )?;
+                    registry.register(Box::new($metric.clone()))?;
+                )*
+                Ok(Self { registry, $($metric,)* })
+            }
```

Comment:
```suggestion
                #[doc = $help:expr]
                $(#[$meta:meta])*
                $metric_vis:vis $metric:ident: $ty:ty $([$($label:expr),* $(,)?])?
            ),* $(,)?
        }
        namespace: $namespace:expr,
    ) => {
        $vis struct $name {
            registry: prometheus::Registry,
            $(
                $(#[$meta])*
                $metric_vis $metric: $ty,
            )*
        }
        impl $name {
            $vis fn new() -> Result<Self, prometheus::Error> {
                let registry = prometheus::Registry::new();
                $(
                    $(#[$meta])*
                    {
                        let $metric = <$ty>::from_opts(
                            prometheus::Opts::new(stringify!($metric), $help)
                                .namespace($namespace)
                                $(.variable_labels(vec![$($label.into()),*]))?
                        )?;
                        registry.register(Box::new($metric.clone()))?;
                    }
                )*
                Ok(Self { registry, $($metric,)* })
            }
```

Reply:
Note that this is not correct: you can't use a block here because then the `let $metric` is scoped only inside the block, and is not available in the constructor below. Applying the attribute on each statement works though.